src/main/java/edu/uw/edm/docfinity/client/DocFinityClient.java

```diff
@@ -0,0 +1,273 @@
+        executeArgs.setDocumentId(documentId);
+        executeArgs.setDocumentTypeId(documentTypeId);
+        executeArgs.setDocumentTypeName(args.getDocumentType());
+        executeArgs.setCategory(args.getCategory());
+        executeArgs.setClientFields(args.getMetadata());
+        executeArgs.setMetadataMap(metadata);
+        executor.executeDatasources(executeArgs).stream().forEach(field -> builder.addValue(field));
+
+        // 3. Index and commit the document using the calculated values from datasources.
+        builder.validateAllRequiredFieldsHaveValue();
+        List<DocumentIndexingMetadataDTO> indexingDtos = builder.build();
+        DocumentIndexingDTO indexingDto =
+                new DocumentIndexingDTO(documentTypeId, documentId, indexingDtos);
+
+        DocumentIndexingDTO indexedDto =
+                this.service.indexDocuments(indexingDto).stream().findFirst().get();
+        return buildIndexResult(args, metadata.values(), indexedDto);
+    }
+
+    /**
+    * Reindexes a document to DocFinity.
+    *
+    * @param args Class that encapsulates arguments for reindex document operation.
+    */
+    public IndexDocumentResult reindexDocument(IndexDocumentArgs args) throws Exception {
+        Preconditions.checkNotNull(args, "args is required.");
+        args.validate();
+
+        String documentId = args.getDocumentId();
+
+        // 1. Get the document type id from the category and document names.
+        String documentTypeId = getDocumentTypeId(args.getCategory(), args.getDocumentType());
+        log.info("Retrieved document type id: {}", documentTypeId);
+
+        // 2. Get all metadata prompts and validate inputs
+        Map<String, MetadataDTO> metadata = getDocumentMetadataMap(documentTypeId, documentId);
+        DocumentIndexingDTO indexingData = service.getDocumentIndexingData(documentId);
+
+        IndexingMetadataBuilder builder =
+                new IndexingMetadataBuilder(
+                                args.getDocumentType(), metadata, indexingData.getIndexingMetadata())
+                        .addValues(args.getMetadata());
+
+        // 3. Execute datasources.
+        DatasourceExecutor executor = new DatasourceExecutor(this.service);
+        ExecuteDatasourceArgs executeArgs = new ExecuteDatasourceArgs();
+        executeArgs.setDocumentId(documentId);
+        executeArgs.setDocumentTypeId(documentTypeId);
+        executeArgs.setDocumentTypeName(args.getDocumentType());
+        executeArgs.setCategory(args.getCategory());
+        executeArgs.setClientFields(args.getMetadata());
+        executeArgs.setMetadataMap(metadata);
+        executor.executeDatasources(executeArgs).stream().forEach(field -> builder.addValue(field));
+
+        // 4. Reindex the document using the calculated values from datasources.
+        builder.validateRequiredFieldsPresentHaveValue();
+        List<DocumentIndexingMetadataDTO> indexingDtos = builder.build();
```

Comment:
Consider `metadataDtos` to reduce confusion, currently `indexingDtos` and `indexingDto` are different types.
```suggestion
        List<DocumentIndexingMetadataDTO> metadataDtos = builder.build();
```